File: 기존 알고리즘 풀이/CLASS 3/Baekjoon2579.cpp
#include <iostream>
using namespace std;

int max(int a, int b)
{
    return a > b ? a : b;
}
int main()
{

    int n;

    cin >> n;

    int num[n];
    int dp[n];
    for (int i = 0; i < n; i++)
    {
        cin >> num[i];
    }
    int count = 0;
    int ans = 0;

    dp[0] = num[0];
    dp[1] = max(num[0], num[0] + num[1]);
    dp[2] = max(dp[0] + num[2], num[1] + num[2]);

    for (int i = 3; i < n; i++)
    {
        dp[i] = max(dp[i - 2] + num[i], dp[i - 3] + num[i - 1] + num[i]);
    }
    cout << dp[n - 1];
}

File: 기존 알고리즘 풀이/Baekjoon2579.py

import sys

N=int(sys.stdin.readline())


global count
count=0
dep=[0]*(N+1)
ans=[]
for i in range(0,N):
    A=int(sys.stdin.readline())
    ans.append(A)

dep[N]=ans[N-1]


icount=0

for i in range(N-1,-1,-1):
    if(i+1==N):
        dep[i]=ans[i-1]+dep[i+1]
    else:
        if(i==0):
            if(max(dep[i+1],dep[i+2])==dep[i+1]):
                icount=icount+1
            dep[i]=max(dep[i+1],dep[i+2])
        else:
            if(max(dep[i+1],dep[i+2])==dep[i+1]):
                if(icount+1==3 or (i+(3-icount+1)==N)):
                    dep[i]=ans[i-1]+dep[i+2]
                    icount=0
                else:
                    icount=icount+1
                    dep[i]=ans[i-1]+max(dep[i+1],dep[i+2])
            else:
                dep[i]=ans[i-1]+max(dep[i+1],dep[i+2])
    
print(dep[1])